bftengine/src/bftengine/messages/ReplicaAsksToLeaveView.cpp

```diff
@@ -0,0 +1,66 @@
+// Concord
+//
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").  You may not use this product except in
+// compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include <cstring>
+
+#include <bftengine/ClientMsgs.hpp>
+#include "OpenTracing.hpp"
+#include "ReplicaAsksToLeaveView.hpp"
+#include "SysConsts.hpp"
+#include "Crypto.hpp"
+#include "ViewsManager.hpp"
+
+namespace bftEngine {
+namespace impl {
+
+ReplicaAsksToLeaveView::ReplicaAsksToLeaveView(
+    ReplicaId srcReplicaId, ViewNum v, Reason r, uint16_t sigLen, const concordUtils::SpanContext& spanContext)
+    : MessageBase(srcReplicaId,
+                  MsgCode::ReplicaAsksToLeaveView,
+                  spanContext.data().size(),
+                  sizeof(Header) + spanContext.data().size() + sigLen) {
+  b()->genReplicaId = srcReplicaId;
+  b()->viewNum = v;
+  b()->reason = (uint8_t)r;
+  b()->sigLength = sigLen;
+  std::memcpy(body() + sizeof(Header), spanContext.data().data(), spanContext.data().size());
+}
+
+ReplicaAsksToLeaveView* ReplicaAsksToLeaveView::create(ReplicaId senderId,
+                                                       ViewNum v,
+                                                       Reason r,
+                                                       const concordUtils::SpanContext& spanContext) {
+  const size_t sigLen = ViewsManager::sigManager_->getMySigLength();
+
+  ReplicaAsksToLeaveView* m = new ReplicaAsksToLeaveView(senderId, v, r, sigLen, spanContext);
+
+  auto position = m->body() + sizeof(Header);
+  std::memcpy(position, spanContext.data().data(), spanContext.data().size());
+  position += spanContext.data().size();
+
+  ViewsManager::sigManager_->sign(m->body(), sizeof(Header) + spanContext.data().size(), position, sigLen);
+
+  return m;
+}
+
+void ReplicaAsksToLeaveView::validate(const ReplicasInfo& repInfo) const {
+  ReplicaId idOfGeneratedReplica = b()->genReplicaId;
```

Comment:
Shouldn't we verify the header size before start referencing it?

Reply:
Yes, definitely, thanks @yuliasherman :)